### PR TITLE
Appconfig endpoint

### PR DESCRIPTION
**Pull request by nickvergessen**
- Required for #1447 

@LukasReschke @MorrisJobke

### CODE REVIEW

**Comment by MorrisJobke** [comment]
* [ ] @nickvergessen mumbled something about "I wanted to move this somehow in the provisioning API" ...

**Comment by MorrisJobke** [comment]
@nickvergessen Fails ;)

```
PHP Fatal error:  Call to a member function getUpdater() on null in /drone/src/github.com/nextcloud/server/lib/private/Files/View.php on line 695
```

**Comment by nickvergessen** [comment]
@icewind1991

**Comment by icewind1991** [comment]
I would prefer to return a promise from `OCP.AppConfig.*` instead of passing callbacks

**Comment by nickvergessen** [comment]
Can you just take care of fixing the failing phpunit tests which I didn't touch?

We can discuss promises later, although I don't feel a lot like breaking the current js API.

**Comment by icewind1991** [comment]
> PHP Warning:  opendir(/drone/src/github.com/nextcloud/server/data-autotest): failed to open dir: No such file or directory in /drone/src/github.com/nextcloud/server/tests/lib/TestCase.php on line 300

Drone messing up?

**Comment by MorrisJobke** [comment]
> Drone messing up?

I restarted it. Let's see

**Comment by MorrisJobke** [comment]
Rebased to trigger CI

**Comment by nickvergessen** [comment]
@icewind1991 
> PHP Fatal error: Call to a member function getUpdater() on null in /drone/src/github.com/nextcloud/server/lib/private/Files/View.php on line 695

why are file tests always failing. :(

**Comment by nickvergessen** [comment]
I think I found the issue, will fix myself

**Comment by nickvergessen** [comment]
Rebased for #1822 and added php unit tests 

@icewind1991 do you want to do the promise stuff in here, or can we do it in a different PR? I have no idea what that is and how it works.

File: core/js/public/appconfig.js [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+	 */
+	_call: function(method, endpoint, options) {
+
+		$.ajax({
+			type: method.toUpperCase(),
+			url: OC.linkToOCS('apps/provisioning_api/api/v1', 2) + 'config/apps' + endpoint,
+			data: options.data || {},
+			success: options.success,
+			error: options.error
+		})
```
Comment (MorrisJobke) [inline core/js/public/appconfig.js]: Semicolon

File: core/js/public/appconfig.js [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+
+		this._call('post', '/' + app + '/' + key, options);
+	},
+
+	/**
+	 * @param {string} app
+	 * @param {string} key
+	 * @param {Object} [options]
+	 * @param {function} [options.success]
+	 * @param {function} [options.error]
```
Comment (MorrisJobke) [inline core/js/public/appconfig.js]: Proper documentation for the parameters would be nice

File: core/js/public/appconfig.js (hunk 1 above)
Comment (nickvergessen) [inline core/js/public/appconfig.js]: Fixed

File: core/js/public/appconfig.js (hunk 2 above)
Comment (nickvergessen) [inline core/js/public/appconfig.js]: Well no documentation means nothing you should use?
Feel free to add something if you think it should be.

File: apps/provisioning_api/appinfo/routes.php [hunk 3]
```diff
@@ -61,5 +56,11 @@
 		['root' => '/cloud', 'name' => 'Users#addSubAdmin', 'url' => '/users/{userId}/subadmins', 'verb' => 'POST'],
 		['root' => '/cloud', 'name' => 'Users#removeSubAdmin', 'url' => '/users/{userId}/subadmins', 'verb' => 'DELETE'],
 
+		// Config
+		['name' => 'AppConfig#getApps', 'url' => '/api/v1/config/apps', 'verb' => 'GET'],
+		['name' => 'AppConfig#getKeys', 'url' => '/api/v1/config/apps/{app}', 'verb' => 'GET'],
+		['name' => 'AppConfig#getValue', 'url' => '/api/v1/config/apps/{app}/{key}', 'verb' => 'GET'],
+		['name' => 'AppConfig#setValue', 'url' => '/api/v1/config/apps/{app}/{key}', 'verb' => 'POST'],
+		['name' => 'AppConfig#deleteKey', 'url' => '/api/v1/config/apps/{app}/{key}', 'verb' => 'DELETE'],
```
Comment (MorrisJobke) [inline apps/provisioning_api/appinfo/routes.php]: Do we really want to name this `/api/` again?

This then results in stuff like `http://localhost/server/ocs/v2.php/apps/provisioning_api/api/v1/config/apps/files_sharing/outgoing_server2server_share_enabled`

**Comment by MorrisJobke** [comment]
Beside the naming comment this is fine by me and works 👍

File: apps/provisioning_api/appinfo/routes.php (hunk 3 above)
Comment (nickvergessen) [inline apps/provisioning_api/appinfo/routes.php]: this is what all ocs routes use

**Comment by rullzer** [comment]
LGTM

**Pull request merged**